openforcefield/typing/engines/smirnoff/parameters.py

```diff
@@ -488,9 +495,41 @@ def to_dict(self, discard_cosmetic_attributes=True):
                     smirnoff_dict[attrib_name + str(idx+1)] = val
             else:
                 smirnoff_dict[attrib_name] = attrib_value
+            if attrib_name == 'smirks':
+                pass
 
         return smirnoff_dict
 
+    def add_cosmetic_attribute(self, attr_name, attr_value):
+        """
+        Add a cosmetic attribute to this ParameterType object. This attribute will not have a functional effect
+        on the object in the Open Force Field toolkit, but can be written out during output.
+
+        Parameters
+        ----------
+        attr_name : str
+            Name of the attribute to define for this ParameterType object.
+        attr_value : str
+            The value of the attribute to define for this ParameterType object.
+        """
+        setattr(self, attr_name, attr_value)
```

Comment:
I'm not 100% sure we want to expose the cosmetic attributes as class attributes. In XML, for example, `my-attr-name` is a perfectly legal attribute, but it's an illegal Python variable name. Maybe we just want to expose a set `ParameterHandler.cosmetic_attributes` that can host only strings that the users can modify how they want. We can create a special class of set if we want to do some validation.

Reply:
> I'm not 100% sure we want to expose the cosmetic attributes as class attributes.

Maybe pedantic, but I think this implementation deals with the attributes on the instance level, and not the class level. I'd love to be corrected if I'm wrong about this.

> In XML, for example, my-attr-name is a perfectly legal attribute, but it's an illegal Python variable name.

This is a really interesting point. Thinking down this road, I wonder what kinds of trouble we might encounter when trying to read/write in other serialization formats. Can they all support the same set of attribute name strings, or is there a possibility that a name that's valid in XML won't be valid in, say, BSON? 

I suspect that Python attribute names will be the most restrictive standard -- that is, any attribute name that is valid for access as `ParameterType.attribute_name` can be serialized in any of our planned serialization formats. 

> Maybe we just want to expose a set ParameterHandler.cosmetic_attributes that can host only strings that the users can modify how they want. We can create a special class of set if we want to do some validation.

If I understand correctly, you're proposing that each `ParameterHandler`/`ParameterType` instance store a `dict` (or other mapping) called `self._cosmetic_attributes` to hold all of its key:value pairs. This would allow us to use a wider range of attributes, with names that would otherwise be invalid (like `my-attr-name`). A tradeoff we would make in this scheme is that users would need to access these attributes using `parameter.cosmetic_attributes['attrname']`, which is a bit clunkier than `parameter.attrname`

Maybe we could get the best of both worlds, by overriding the `ParameterType`/`ParameterHandler` `__getattr__` and `__setattr__` functions to first check whether actual attrs with the desired name exist, and if not, checking the `cosmetic_attribute` dict. That way, we'd provide access similar to what we have now, but also let users use otherwise-invalid attribute names by accessing the `cosmetic_attribute` dict directly.

This seems like a good direction, but I'd like to flesh it out more before we move forward. I'll turn this comment into an Issue so we can aim for it in a subsequent milestone. Until then, the current implementation should be _more_ restrictive, so we wouldn't need to break the API to transition to a more flexible scheme later.

Reply:
Opened #338 to discuss this further.

@andrrizzi, do you think it's safe to add this method to the API, or do you think it will permit a behavior that we will prohibit in the future?

Reply:
> I'd love to be corrected if I'm wrong about this.

You're totally right. That was poor choice of words on my part.

> Opened #338 to discuss this further.

Commented!

> do you think it's safe to add this method to the API, or do you think it will permit a behavior that we will prohibit in the future?

If we are 100% certain we'll support the dictionary syntax in the future before releasing 0.4, we might as well remove these methods to avoid breaking the API later. But if we still have doubts we can definitely keep them around until we make a decision (maybe adding a warning in the docs).

Reply:
We should continue the discussion, but I think I'll keep `add`/`delete_cosmetic_attribute` in there for the current release. When the 0.3.0 toolkit loads cosmetic attributes, they do become accessible via `ParameterType.jeffs_cosmetic_attr`, so moving to a dict would break that API.